### PR TITLE
Prefix cache fixes

### PR DESCRIPTION
**Pull request by wdykas**
- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
• - Preserve newer prefix-cache mappings when an older block with the same hash is released or evicted.
  - Apply the ownership check consistently to KV and Mamba caches.
  - Always invalidate removed Mamba slots.
  - Add regression tests for duplicate-hash release and eviction.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by wdykas** [comment]
@lmcafee-nvidia plz take a look when your free

File: megatron/core/inference/contexts/kv_block_allocator.py [hunk 1]
```diff
@@ -403,17 +403,22 @@ def _deregister_blocks(self, block_ids: Tensor) -> None:
...
-            maxlen=0,
-        )
-
-        # Notify Mamba slot allocator (if wired) to clean up its state
+        block_ids_list = block_ids.tolist()
+        block_id_set = set(block_ids_list)
+        candidate_hashes = set(hashes) - {-1}
+
+        # A hash can be re-registered to a newer physical block before an older
+        # duplicate is released. Remove only mappings still owned by these blocks.
```
Comment (santhnm2) [inline megatron/core/inference/contexts/kv_block_allocator.py:411]: Is this still true in main? I pushed a fix to [!5995](https://github.com/NVIDIA/Megatron-LM/pull/5995) (commit [`73fdd5db8b1d936e23b43cc5bcbd706c5bcf487a`](https://github.com/NVIDIA/Megatron-LM/pull/5995/changes/73fdd5db8b1d936e23b43cc5bcbd706c5bcf487a)) to prevent these duplicate registrations

Comment (wdykas) [reply megatron/core/inference/contexts/kv_block_allocator.py:411]: I think that MR prevents the same (block_id, hash) pair from being registered twice. This PR handles a different case I think... could be for sure missing something.
 2 different physical blocks registered with the same hash. I think on main it still allows the newer block to replace the hash-map entry, after which releasing the older block  removes the newer block’s mapping. This tries to make KV cleanup verify ownership before removing the mapping.

Comment (santhnm2) [reply megatron/core/inference/contexts/kv_block_allocator.py:411]: So the `test_mamba_duplicate_hash_cleanup_preserves_new_mapping` test is testing this by registering block 0 with hash 101 and then registering block 1 with the same hash 101:
```
allocator.register_block_hashes_batch([0], [101])
allocator.register_block_hashes_batch([1], [101])
```
But I don't think that should ever be happening in practice, and in `register_kv_block_hashes` this case would throw an explicit error here: 
https://github.com/NVIDIA/Megatron-LM/blob/708f0a5eaa7b71b89997b8d1dd2222bcbea187a4/megatron/core/inference/contexts/kv_block_allocator.py#L337-L345

Comment (santhnm2) [reply megatron/core/inference/contexts/kv_block_allocator.py:411]: Although I guess this is checking if the same block is being registered with different hashes rather than two different blocks being registered with the same hash - both cases seem like they should be invalid though
